### PR TITLE
added support for forcing images to be rebuilt when ./sandbox clean i…

### PR DESCRIPTION
**Pull request by overhead525**
…s called

### CODE REVIEW

**Comment by overhead525** [comment]
I need to also add a commit to explain this in the README.md

However, its functionality goes hand in hand with PR #50, so let's just leave this open until we've got that one merged. Then i'll squash in another commit to update the README.md file on this PR.

**Comment by Mine77** [comment]
@fabrice102  Could you please review this bounty and provide comments to our contributor? Thank you.

File: sandbox [hunk 1]
```diff
@@ -169,6 +169,25 @@ sandbox () {
     docker-compose -f "$SANDBOX_DIR/docker-compose.yml" "$@"
   }
 
+  dc_no_cache () {
+    docker-compose build --no-cache
```
Comment (fabrice102) [inline sandbox:173]: You should always use `dc` instead of `docker-compose` because `dc` specify the docker-compose file

Comment (fabrice102) [inline sandbox:173]: Use `dc` instead of `docker-compose`, as `dc` specifies correctly the `docker-compose.yml` file

File: sandbox [hunk 2]
```diff
@@ -169,6 +169,25 @@ sandbox () {
...
+    dc "$@"
+  }
+
+  dc_with_clean_check () {
+    CLEAN_FILE=.clean
```
Comment (fabrice102) [inline sandbox:178]: sandbox may not be run in the correct folder, the file must be

`CLEAN_FILE="$SANDBOX_DIR/.clean"`

File: sandbox [hunk 3]
```diff
@@ -273,6 +292,7 @@ sandbox () {
     echo "* docker rmi $(docker images -f "dangling=true" -q)"
     docker rmi $(docker images -f "dangling=true" -q)          || true
     rm "$ACTIVE_CONFIG_FILE" > /dev/null 2>&1                  || true
+    echo 'clean function executed' > .clean;
```
Comment (fabrice102) [inline sandbox:295]: Same issue as above.
Simplest is to create a variable `CLEAN_FILE` that is global.

**Review by fabrice102 — CHANGES_REQUESTED** [review]
@overhead525 Thanks for the commit, can you make it so that `* docker-compose build --no-cache` is printed on screen or in LOG before `* docker-compose up -d`.

This would mean moving a bit the logic.
Maybe have instead a function `dc_rebuild_if_needed`, that would test if rebuilding is needed like `dc_build_with_clean_check` and then print `* docker-compose build --no-cache` (with potentially the sentence `".clean file found ..."`) and run `dc build --no-cache`.

You would then call this at line 393, 396, and 402.
Line 396, you call it with `>> "$SANDBOX_LOG"`
Lines 393 and 402, you simply call it.

Check the effect of the spinner

**Comment by overhead525** [comment]
> @overhead525 Thanks for the commit, can you make it so that `* docker-compose build --no-cache` is printed on screen or in LOG before `* docker-compose up -d`.
> 
> This would mean moving a bit the logic.
> Maybe have instead a function `dc_rebuild_if_needed`, that would test if rebuilding is needed like `dc_build_with_clean_check` and then print `* docker-compose build --no-cache` (with potentially the sentence `".clean file found ..."`) and run `dc build --no-cache`.
> 
> You would then call this at line 393, 396, and 402.
> Line 396, you call it with `>> "$SANDBOX_LOG"`
> Lines 393 and 402, you simply call it.
> 
> Check the effect of the spinner

I see. So currently the function `dc_build_with_clean_check` handles the checking if rebuilding is necessary AND the calling of the `dc up` command. Your request requires them to be separated. Gotcha.

File: sandbox [hunk 4]
```diff
@@ -373,12 +382,28 @@ sandbox () {
         dc up
```
Comment (overhead525) [inline sandbox:382]: Inside of this `if` conditional, we only have `dc up`. I wasn't sure if you guys wanted to run this new logic when the user tries **interactive mode** or not.

**Comment by overhead525** [comment]
https://github.com/algorand/sandbox/blob/df654aafecaa48f1c5371094147cd4999a74c122/sandbox#L381-L382

Inside of this `if` conditional block, we only have `dc up`. I wasn't sure if we wanted to have the rebuilding images logic if the user is in **interactive mode**. It's up to you guys.

**Comment by fabrice102** [comment]
@overhead525 I've made a PR to your PR with cleaner code. Please merge in your PR.

The main point of the PR is to show you that you can create a function to avoid duplicating codes.
Duplicated code usually quickly creates issues.
In particular in your code, you had a `>> "$SANDBOX_LOG` where it should not be.

**Comment by overhead525** [comment]
> @overhead525 I've made a PR to your PR with cleaner code. Please merge in your PR.
> 
> The main point of the PR is to show you that you can create a function to avoid duplicating codes.
> Duplicated code usually quickly creates issues.
> In particular in your code, you had a `>> "$SANDBOX_LOG` where it should not be.

Just merged. Thanks for pointing that out for me. Having a `>> $SANDBOX_LOG` is indeed indicative of the issue.

**Comment by winder** [comment]
I think `>> $SANDBOX_LOG` is still wrong, if you run with `-i` you wouldn't see the rebuild output.

File: sandbox [hunk 5]
```diff
@@ -169,6 +170,17 @@ sandbox () {
     docker-compose -f "$SANDBOX_DIR/docker-compose.yml" "$@"
   }
 
+  rebuild_if_needed () {
+    if [ -f "$CLEAN_FILE" ]; then
+      rm "$CLEAN_FILE"
+      echo ".clean file found in sandbox directory. Rebuilding images..."
+      echo "* docker-compose build --no-cache"
+      dc build --no-cache             >> "$SANDBOX_LOG"
```
Comment (fabrice102) [inline sandbox:178]:
```suggestion
      dc build --no-cache
```

**Comment by fabrice102** [comment]
@winder indeed, it needs to be removed. Made a suggestion above.

**Pull request merged**